zorg/buildbot/process/buildrequest.py

```diff
@@ -22,14 +22,20 @@ def collapseRequests(master, builder, req1, req2):
         ('buildsets', str(req2['buildsetid'])))
 
     # Fetch the buildset properties.
-    selfBuildsetPoperties = yield \
-        master.db.buildsets.getBuildsetProperties(
-            str(req1['buildsetid'])
-        )
-    otherBuildsetPoperties = yield \
-        master.db.buildsets.getBuildsetProperties(
-            str(req2['buildsetid'])
-        )
+    selfBuildsetPoperties = yield master.db.buildsets.getBuildsetProperties(
+        str(req1["buildsetid"])
+    )
+    otherBuildsetPoperties = yield master.db.buildsets.getBuildsetProperties(
+        str(req2["buildsetid"])
+    )
+
+    # If the build is going to be a clean build anyway, we can collapse a clean
+    # build and a non-clean build.
+    if getattr(builder.config.factory, "clean"):
```

Comment:
Change to `if getattr(builder.config.factory, "clean", False):`, then you can keep zorg/buildbot/process/factory.py unchanged.

Reply:
I've fixed the default, I mistakenly thought it returned None like a dictionary .get would.

I don't see how that means I don't have to change factory.py though. Without that change no factory would have the clean attribute set on it.

Unless you mean I should split that change out so that this PR is purely adding the ability to collapse more, then a follow up enables it?

(which is a decent idea, it's still possible I'm breaking unrelated builders here)

Reply:
Unless `clang` is not specified as the parameter explicitly the following code in line 87 do the trick:
```
        # Preserve the rest of the given extra attributes if any, so we could
        # expand the factory later.
        for k,v in kwargs.items():
            setattr(self, k, v)
```

Reply:
I see what you mean now, I didn't see that loop.

It looks like things we know we want to use are popped from kwargs and then that loop catches the remaining ones. So I've changed clean to pop from kwargs with a default False.

We could rely on that loop entirely but it's pretty cryptic. An explicit check implies it's something we care about and is a target for anyone searching for this later.